data/src/main/java/com/linkedin/data/schema/PathSpecSet.java

```diff
@@ -0,0 +1,440 @@
+/*
+   Copyright (c) 2012 LinkedIn Corp.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package com.linkedin.data.schema;
+
+import com.linkedin.data.template.RecordTemplate;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.lang3.builder.EqualsBuilder;
+import org.apache.commons.lang3.builder.HashCodeBuilder;
+
+
+/**
+ * Represents an immutable set of {@link PathSpec}s.
+ *
+ * <p>Path spec set is a convenient wrapper for a collection of {@link PathSpec}. A few advantages of using this class
+ * over manually passing around {@code Set<PathSpec>} is:
+ *
+ * <ul>
+ *   <li>guaranteed immutable representation</li>
+ *   <li>
+ *     explicitly supports "full projection" which translates into fetch everything (default for rest.li resources),
+ *     as well as "empty" which translates into fetch no fields
+ *   </li>
+ *   <li>mutable builder for incrementally assembling an immutable {@link PathSpecSet}</li>
+ *   <li>built-in utility to mask a {@link RecordTemplate} with the stored path specs</li>
+ * </ul>
+ *
+ * @author Joseph Florencio
+ */
+final public class PathSpecSet
+{
+  private final static PathSpecSet EMPTY = new PathSpecSet(Collections.emptySet(), false);
+  private final static PathSpecSet FULL_PROJECTION = new PathSpecSet(Collections.emptySet(), true);
+
+  private final Set<PathSpec> _pathSpecs;
+  private final boolean _fullProjection;
```

Comment:
I'm skeptical about using the terminology "full projection", as PathSpecs are not inherently tied to Rest.li projections (that's just one application). However, one could make the argument that the term "projection" is a generic term and isn't implying that it's used for _Rest.li_ projections. What do you think?

I'm also curious to know how a "full projection" PathSpecSet is used. Do you have an example of one use case?

Reply:
maybe we can change to name "universal". Otherwise the "full" return a empty set and NULL in toArray seems unnatural to me as well

Reply:
In Rest.li world, when you have a PathSpecSet you may have logic that is skipped if the projection doesn't include a specific field:
```
if (pathSpecSet.contains(MyRecord.fields().subrecord()) {
    // do a bunch of stuff that is costly but unnecessary if the field subrecord isn't requested
}
```

And then given the ResourceContext, you can easily build a PathSpecSet with `PathSpecSet.of(resourceContext.getProjectionMask())`. 

When no projection was specified, the ProjectionMask is null and you want to make sure that `contains(PathSpec)` returns true for any PathSpec and that's the meaning of a fullProjection.

Reply:
For the `toArray()` returning `null`, this is consistent with the behavior of the method `fields(PathSpec[])` in the Rest.li builders.

Reply:
I'm used to full projection so it is difficult for me to come up with an alternative. I'm not sure universal as an adjective works but I could see `universe` or `universeSet` as suitable. Here are some alternatives that I found in a thesaurus dictionary:
- globalSet
- allInclusiveSet
- blanketSet

Reply:
> When no projection is specified, the ProjectionMask is null and you want to make sure that `contains(PathSpec)` returns true for any PathSpec and that's the meaning of a fullProjection.

In this specific use case it makes sense, but this object is simply a set of path specs. An empty path spec set articulates to a full projection when it's used in the specific application of projection, but path specs aren't only for projections. What does an empty path spec set mean when it's being used for indexing values from a complex data object, or matching ReadOnly/CreateOnly predicates? I don't necessarily think that an empty path spec set contains everything for all use cases.

Reply:
thanks for the explanation, I will vote for "allInclusiveSet".

Reply:
Synced up with @GuiForget about this offline. He clarified that an empty set is not necessarily the same as an "all inclusive" or "full projection" set, which must be created explicitly as such.

As for naming, I say we avoid using the terminology "full projection", and instead use something that implies inclusivity. `allInclusive`/`isAllInclusive` works for me. `complete` may also be good, but I'm not sure if that has a different technical definition when it comes to sets.

Reply:
I renamed fullProjection to allInclusive and propagated the changes throughout the classes and unit tests